FINAL

lilumi opened the pull request.

lilumi-li committed:
Здача проекта

lilumi-li committed:
Сдача проекта

lilumi-li committed:
FINAL

slebedev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,124 @@
+- (id)init {
+   if ( self= [super init]) {
+      self.DNA = [NSMutableArray arrayWithCapacity:100];
+       for (int i=0; i<=99; i++) {
```
Ну уж для себя-то могли бы и `for (int i=0; i<100; i++)` написать. Легче же читается. В инициализации с произвольным числом то же самое.

slebedev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,124 @@
+
+- (id)initWithLenght:(int)length {
+    if ( self= [super init]) {
+        self.DNA = [NSMutableArray arrayWithCapacity:length-1];
```
Перед тем, как так делать, нужно проверить, не отрицательное ли length?

lilumi replied:
Ок, сделаю. А как добавить еще один пул-реквест, чтобы не удалять старый? А то у меня при попытке добавить выскакивает такое: http://d.pr/i/PJDA

slebedev replied:
Просто делайте коммит, повторный пулл-реквест делать не нужно, все изменения здесь отобразятся.

Pull Request нужно делать один раз, чтобы "связать" свой проект с этим репозиторием.

lilumi replied:
Спасибо. 
А что мне возвращать, когда length меньше 0 ?

slebedev replied:
На данном этапе хватит того, что вы насильно приравняете в этом случае length к 0. Что-то типа

`if (length<0) length=0;`

lilumi-li committed:
FINAL

lilumi-li committed:
FINAL

lilumi-li committed:
FINAL

функция print заменена на description

lilumi-li committed:
FINAL

улучшил алгоритм формирования случайных индексов для лучшего чтения
кода и оптимизации памяти

aquaxp commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+- (NSString* ) randomCharacter {
+    NSString* arrayOfChars = @"ACGT";
+    int i = arc4random() % 4;
+    NSString *value = [arrayOfChars substringWithRange:NSMakeRange(i, 1)];
```
А не проще взять i-й элемент arrayOfChars, а не вырезать подстроку длиной в один символ? Если так делается для безопасности, то все равно это никак в дальнейшем коде не обрабатывается.

lilumi replied:
Согласен - проще, но на тот момент я не знал как взять этот элемент, что нагуглил, то и использовал :)

lilumi replied:
Кстати, замерил скорость выполения. При больших итерациях (100 млн. раз) мой код выполняется за 10 секунд, а characterAtIndex выполнялся аж 98 секунд. http://d.pr/i/7bgX

aquaxp replied:
Странно... интересно, откуда такой большой оверхэд... Проверьте еще раз, избавившись от повторного приведения к NSString и подступаясь к элементам как arrayOfChars[i], изменив NSString *arrayOfChars[4] = {@"A",@"T",@"G",@"C"};

lilumi replied:
О, при таком коде http://d.pr/i/6t73 второй вариант выполняется за 5 секунд.  Тааак, я чувствую, что надо половину кода переписать. 
Еще хочу потестировать на скорость свой код формирования случайных индексов.

vratushnyi replied:
return [@"ACGT" substringWithRange:NSMakeRange(arc4random() % 4, 1)];

lilumi replied:
@vratushnyi да, это понятно, что если переменная используется только один раз, то не зачем ее обьявлять, но здесь (да и в других местах программы) я заводил такие переменные для того, чтобы в дебагере отслеживать значения потому, что Xcode не показывает значения выражений, значения одного обьекта из массива и т.д.

slebedev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+
+- (id)initWithLenght:(int)length {
+    if ( self = [super init]) {
+        if (length<=0) length=1; //проверка на нулевое и отрицательные значения длины ДНК
```
Вот тут лучше так не делать. Лучше всего "вытолкнуть" ошибку с помощью `@throw` или NSAssert.

lilumi replied:
Я просто не знаю что такое @throw или NSAssert :) 
Как лучше написать, чтобы человек увидел сообщение об ошибке, которое он сможет понять, а не набор цифр ?

lilumi replied:
@slebedev кстати, еще вчера ты про эту строчку говорил так: «На данном этапе хватит того, что вы насильно приравняете в этом случае length к 0. Что-то типа if (length<0) length=0;»

slebedev replied:
@lilumi разве в этом методе я так говорил? Вроде как говорили о методе mutate, нет?

slebedev replied:
@lilumi вижу, получается, я мыслями был в другом методе. =) Простите меня. А про throw и NSAssert стоит почитать - это выталкивание ошибки, теоретически поможет программисту, который будет пользоваться вашим классом.

lilumi replied:
А где порекомендуешь почитать об этом на русском? А то я пробую читать в официальной документации, но мне сложно дается понимание на английском языке.

slebedev replied:
Завтра уже, я спать. ;)

lilumi replied:
В любом случае спасибо, что многое разьяснил. 
Вот смотри, использовал я @throw и в итоге получил такую картину: http://d.pr/i/AzSt - а хотелось бы, чтобы пользователь просто увидел сообщение об ошибке без всяких отчетов с адресами памяти.
А в идеале, чтобы программа ему предложила ввести другое корректное значение.

RSEvseev replied:
Можно использовать [NSException raise: format:] - сгенерируется исключение

lilumi replied:
попробовал if (length<=0) [NSException raise:@"Неверная длина ДНК" format:@"Длина ДНК должна быть положительным не нулевым числом"]; и все равно получаю результат http://d.pr/i/AzSt а хочется пользователю показывать только сообщение об ошибке и ничего больше.

RSEvseev replied:
а вот для этого нужно вызов mutate: запихивать в @try..@catch и обрабатывать исключение.

lilumi replied:
@RSEvseev а почему в mutate ? Мы же сейчас находимся в инициализации. Можно как-то в эту инициализацию запихнуть возможность пользователю вручную ввести значение количества символов в ДНК и после этого сгенерировать эту ДНК ?

RSEvseev replied:
упс... обсуждение длиное, уже забыл, с чего все начиналось, искренне считал, что речь о процентах в mutate: 
в случае инита думаю имеет смысл длину проверять до вызова [super init] и в случае ее некорректного значения возвращать nil - объект не создается, программа не вываливается, все счастливы ))

slebedev replied:
@RSEvseev всё бы хорошо, если бы Objective-C не мог бы посылать сообщения объекту nil. А он может, это в порядке вещей.

@lilumi то, что вы хотите сделать, реализуется вылавливанием ошибки. В методе initWithLength её генирируете, а отлавливать должны при вызове, соответственно там и надо использовать. Но как мне объяснили - плохая практика.
Просили ссылку, но у меня конкретно по теме ничего нет, зато есть "общая" ссылка по Objective-C в целом: http://unclemif.com/public/objective-c-frame.html

RSEvseev replied:
@slebedev при чем тут nil и сообщения? Это обычная практика, если в init если не проходит [super init] - возвращается nil. Просто в данном случае мы даже до [super init] не доходим - ибо незачем при неверных начальных условиях.

lilumi replied:
@RSEvseev я так понял, ты проверял мою работу, а оценки ты уже выставил? А то я не вижу оценки, думаю где ее искать ?

RSEvseev replied:
Да, выставил. Заходишь на страницу описания проекта, справа жмешь на Submission Details. Сверху справа Show Rubric

lilumi replied:
Ага, вижу, спасибо. 
А если я сейчас загружу новую версию кода, то эти все коментарии сотрутся или их где-то можно будет прочесть ?

RSEvseev replied:
Комментарии останутся, просто будут свернуты как outdated.

aquaxp commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+    if ([[self DNA] count] > [[aCell DNA] count]) length = [[aCell DNA] count]; //если первая ДНК длиннее второй, то сравниваем только до длины второй ДНК
+    for (i=0; i<length; i++) {
+        
+        if ([self.DNA objectAtIndex:i]!= [aCell.DNA objectAtIndex:i] ) {
```
Сравнивать NSString'и следует так http://stackoverflow.com/questions/3703554/understanding-nsstring-comparison-in-objective-c
P.S. Сам на это напоролся.

lilumi replied:
Круто. Спасибо, не знал. Но по комментариям на стековерфлоу вижу, что компилятор все равно оптимизирует код и результат будет тот же, что с isEqualToString. Но на будущее учту.

RSEvseev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+- (NSString* ) replaceCharacter: (NSString*) exclude {
+    NSString* arrayOfChars = @"ACGT";
+    arrayOfChars = [arrayOfChars stringByReplacingOccurrencesOfString:exclude withString:@""];
+    int i = arc4random() % 3;
```
Здесь правильнее будет int i = arc4random() % [arrayOfChars length];
В этом случае можно не фильтровать exclude на символы в arrayOfChars. Ну и в качестве "побочного эффекта" можно избавиться от функции randomCharacter, вместо нее вызывая [self replaceCharacter: @" "] (ну, или не @" ", а любой другой символ не из множества {AGCT})
Помимо этого, данную функцию вполне можно сделать функцией класса

lilumi replied:
Я не понял фразу "можно не фильтровать exclude на символы в arrayOfChars" — можешь другими словами обьяснить ?
Про побочный эффект я все понял — ты прав. Но только тогда нужно будет название функции придумать другое, которое бы соответствовало двум разным действиям - генерации случайного и замене. Есть варианты, как назвать ?

А чтобы сделать ее функцией класса я должен поставить знак +, так ? И какой профит будет от этого?

RSEvseev replied:
ну пример - тебе в функцию передали символ, которого нет в множестве {AGCT}, например Q. stringByReplacingOccurrencesOfString: withString: в данном случае ничего не заменит - он просто не найдет такого символа. Соответственно, когда ты сделаешь arc4random() % 3, ты потеряешь символ T т.к. у него индекс 3, а у тебя случайное значение выбирается из 0, 1 и 2, т.е. Т при таком условии никогда не выберется.
другой пример - у тебя параметр NSString \* как-никак, значит, мы можем передать, например, @"CT", или @"GC" - после stringByReplacingOccurrencesOfString: withString: у нас останется всего 2 символа, с индексами 0 и 1, а arc4random %3 по прежнему генерит 0, 1 или 2 - соответственно, имеем выход за границу массива. Если же передать @"AGCT", то у нас вообще останется пустая строка... Правда, в этом случае посыплется и мой вариант - arc4random() % 0 не есть хорошо ))) Именно поэтому удобнее использовать не NSString для arrayOfChars, а именно NSArray - там подобная ситуация исключена. Загляни в мой код - там сделано именно так. Я эту функцию обозвал + (NSString *) randomElementWithout: (NSString *) element;

Да, функция класса - плюс вместо минуса. Ну, конкретно в данном задании профита от этого никакого (разве что проверять ее работоспособность без создания экземпляра класса), но по логике это именно функция класса - от членов класса она не зависит и может быть вызвана, когда никакого экземпляра класса не существует.

RSEvseev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+}
+
+- (NSString *) description {
+    return [NSString stringWithFormat:@"%@", [_DNA componentsJoinedByString:@""]];
```
componentsJoinedByString возвращает объект NSString, так что можно сделать просто return [_DNA componentsJoinedByString:@" "];

lilumi replied:
Спасибо! Это все от незнания какие методы есть у классов, какие когда лучше использовать, я как слепой котенок, который натыкается на все подряд — что нашел первым в гугле, то и использовал.

RSEvseev replied:
Рад помочь ))

RSEvseev commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+        if (length<=0) length=1; //проверка на нулевое и отрицательные значения длины ДНК
+            self.DNA = [NSMutableArray arrayWithCapacity:length];
+            for (int i=0; i<length; i++) {
+                [self.DNA insertObject:[self randomCharacter] atIndex:i];
```
Вместо [self.DNA insertObject: atIndex:] в данном случае эффективнее использовать [_DNA addObject:]

lilumi replied:
а почему именно здесь эффективнее использовать _DNA ? в коде других людей, я видел комментарии, когда писали, что лучше использовать self.DNA вместо _DNA так как это вызывает сеттер (или геттер, я запутался).

RSEvseev replied:
геттер - когда мы читаем значение, сеттер - когда записываем. 
в принципе, в твоем случае в части self.DNA vs _DNA разницы практически нет - ты не делал свойство DNA readonly и не защищал его от модификации извне (не переписывал геттер, чтобы он возвращал неизменяемый массив). Если бы ты это сделал - у тебя self.DNA вернул бы не NSMutableArray, а NSArray, который не имеет ни insertObject: atIndex:, ни addObject. Так что здесь выигрышь только от мизерной экономии процессорного времени - не нужно вызывать геттер и возвращаться из него. Но я в комментарии выше акцент делал не столько на _DNA вместо self.DNA, сколько на addObject: вместо insertObject: atIndex: - опять же, он быстрее выполняется, поскольку ему не надо искать индекс i и он просто добавляет объект в конец массива

vratushnyi replied:
Вы отправляете сообщения экземпляру класса (self.DNA или [self DNA]) до завершения метода init. Это плохая идея.
init должен позаботиться о правильной инициализации состояния класса, без использования деталей реализации.
Более детально http://stackoverflow.com/questions/5932677/initializing-a-property-dot-notation

lilumi replied:
Если честно, то я не понял в чем здесь проблема - строчкой выше прошел super init значит обьект создан, память выделена, можно уже с ним работать. 
Ты же в своем коде написал [self DNA] но, насколько я понял синтаксис obj-c, то это идентично self.DNA

vratushnyi replied:
Проблемы нет. Это скорее "плохая привычка". В других случаях getter свойства может что-то делать, кроме как отдавать ссылку на _DNA, т.е. будут side effects (работа с другими переменными и т.д.)

lilumi replied:
А как правильно нужно было сделать? обращаться к prorerty напрямую: [_DNA insertObject:[self randomCharacter] atIndex:i]; ?

vratushnyi replied:
Да. _DNA - переменная, описывающая состояние объекта. Соответствующий getter - уже реализация.

vratushnyi commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+       
+   }
+    
+    return self;
```
В вашем случае реализацию init можно заменить на
- (id)init {
  return [self initWithLenght:100];
  }
  http://developer.apple.com/library/ios/#documentation/general/conceptual/DevPedia-CocoaCore/MultipleInitializers.html

vratushnyi commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+- (int) hammingDistance:(Cell *)aCell {
+    int ham = 0, i;
+    NSInteger length = [[self DNA] count];
+    if ([[self DNA] count] > [[aCell DNA] count]) length = [[aCell DNA] count]; //если первая ДНК длиннее второй, то сравниваем только до длины второй ДНК
```
int length = MIN([self.DNA count], [aCell.DNA count]);

И сразу все понятно без комментариев

lilumi replied:
Как я уже неоднократно говорил - я учу Objective C с нуля и у меня есть только те знания, которые я узнал из лекций на этом сайте и из того, что я нагуглил. Про существовании метода MIN я не знал. 
Спасибо, на будущее буду знать.

vratushnyi replied:
У вас корректная программа. Я стараюсь не просто выставить оценку но и что-то предложить. Относитесь к комментариям (к моим как минимум) как к предложениям, способу поделиться знаниями.

lilumi replied:
Я так и отношусь. И очень благодарен за ценные комментарии. У меня уже готова совсем другая версия программы основанная на всех замечаниях и комментариях от других студентов, но я ее не выкладывал еще, так как хотел, чтобы оценку мне поставили за то, что я написал сам без помощи других.

vratushnyi commented on Cell+Mutator.m:
```diff
@@ -0,0 +1,118 @@
+                i++;
+            }
+            
+        } while (i<n); // пока не сформируем масив с нужным кол-вом индексов
```
При 100 элементах в массиве и 100% замене вы можете получить более 1000 итераций в цикле формирования случайных чисел, т.е. эффективность 10N.
Если установить к-во элементов DNA в 100 000 и менять 100%, одна мутация (на моем ноуте) занимает приблизительно 2 минуты 30 секунд.
Попробуйте рассмотреть например такой алгоритм формирования массива
NSInteger indexArray[[self.DNA count]];
for (int i = 0; i < [self.DNA count]; i++) indexArray[i] = i;
for (int i = 0; i < [self.DNA count]; i++) {
     int rnd = arc4random() % [self.DNA count];
     NSInteger swap = indexArray[rnd];
     indexArray[rnd] = indexArray[i];
     indexArray[i] = swap;
}

при 100 000 и 100% мутации работает не более 1 сек., но в этой реализации используется больше памяти. Возможно вы найдете что-то среднее по времени выполнения и использовании памяти.
Все, ушел гулять с ребенком :)

lilumi replied:
да, спасибо, я пару дней назад тоже размышлял над алгоримтической сложностью этого алгоритма и придумал свое решение, которое исключает вообще повторные циклы использования рандома.
        NSInteger l=[_DNA count], i=0;
        NSMutableArray \* indexArray = [NSMutableArray array], \* sortedArray = [NSMutableArray array];
        for (i=0; i<l; i++) {
            [sortedArray addObject:[NSNumber numberWithInteger:i]];
        }

```
    for (i=0; i<n; i++) {
        int pos = arc4random()%l;
        indexArray[i]=sortedArray[pos];
        [sortedArray removeObjectAtIndex:pos];
        l--;

    }
```

lilumi-li committed:
Final v.2

В проект вненесны изменения в соотвествии с обсуждением с коллегами.
Кстати, спасибо всем тем, кто обсуждал код и расширял мои познания в
obj-c.

Переписан алгоритм генерации массива индексов, который исключает
потенциально бесконечные циклы поиска случайного числа.

lilumi-li committed:
Final v.2.0.1
